src/uct/ib/base/ib_device.c

```diff
@@ -408,6 +428,136 @@ ucs_status_t uct_ib_device_port_check(uct_ib_device_t *dev, uint8_t port_num,
+    for (i = 1; i < len; i++) {
+        if (gid_idx_rates[i] < best_priority) {
+            best_priority = gid_idx_rates[i];
+            gid_index = i;
+        }
+    }
+
+    return gid_index;
+}
+#endif
+
+static int uct_device_get_addr_family(union ibv_gid *gid, int gid_index)
+{
+    const struct in6_addr *raw = (struct in6_addr *)gid->raw;
+    char p[128];
+
+    ucs_debug("testing addr_family on gid index %d: %s",
+              gid_index, inet_ntop(AF_INET6, gid, p, sizeof(p)));
+
+    if (((raw->s6_addr32[0] | raw->s6_addr32[1]) |
+            (raw->s6_addr32[2] ^ htonl(0x0000ffff))) == 0UL ||
+           /* IPv4 encoded multicast addresses */
+           (raw->s6_addr32[0] == htonl(0xff0e0000) &&
+            ((raw->s6_addr32[1] | (raw->s6_addr32[2] ^ htonl(0x0000ffff))) == 0UL))) {
+        return AF_INET;
+    } else {
+        return AF_INET6;
+    }
+}
+
+static ucs_status_t uct_ib_device_set_roce_gid_index(uct_ib_device_t *dev,
+                                                     uint8_t port_num,
+                                                     uint8_t *gid_index)
+{
+    int i, gid_tbl_len = uct_ib_device_port_attr(dev, port_num)->gid_tbl_len;
+    ucs_status_t status          = UCS_OK;
+#if HAVE_DECL_IBV_EXP_QUERY_GID_ATTR
+    int prio_idx, found;
+    struct ibv_exp_gid_attr attr;
+    unsigned *gid_idx_rates      = ucs_alloca(gid_tbl_len * sizeof(*gid_idx_rates));
+    int priorities_arr_len       = sizeof(roce_versions_priorities) /
+                                   sizeof(roce_versions_priorities[0]);
+#else
+    union ibv_gid gid;
+#endif
+
+    for (i = 0; i < gid_tbl_len; i++) {
+#if HAVE_DECL_IBV_EXP_QUERY_GID_ATTR
+        attr.comp_mask = IBV_EXP_QUERY_GID_ATTR_TYPE | IBV_EXP_QUERY_GID_ATTR_GID;
+        if (ibv_exp_query_gid_attr(dev->ibv_context, port_num, i, &attr)) {
+            ucs_error("failed to query gid attributes "
+                      "(%s:%d gid_idx %d). %m", uct_ib_device_name(dev), port_num, i);
+            status = UCS_ERR_INVALID_PARAM;
+            goto out;
+        }
+
+        found = 0;
+        for (prio_idx = 0; prio_idx < priorities_arr_len; prio_idx++) {
+            if ((attr.type == roce_versions_priorities[prio_idx].type) &&
+                (uct_device_get_addr_family(&attr.gid ,i)) ==
+                 roce_versions_priorities[prio_idx].address_family) {
+                gid_idx_rates[i] = roce_versions_priorities[prio_idx].priority;
+                found = 1;
+                break;
+            }
+        }
+
+        if (!found) {
+            ucs_error("unknown RoCE gid type (%d) or no matching address family",
+                      attr.type);
+            status = UCS_ERR_INVALID_PARAM;
+            goto out;
+        }
+#else
+        status = uct_ib_device_query_gid(dev, port_num, i, &gid);
+        if (status != UCS_OK) {
+            goto out;
+        }
+
+        /* assume RoCE v1 */
+        if (uct_device_get_addr_family(&gid, i)) {
```

Comment:
what do we check in this `if`? AFAIU, `uct_device_get_addr_family` always returns non-zero value.